cmd/exec.go

```diff
@@ -15,148 +12,83 @@ import (
+		ui.Errorf(errMsg)
+		os.Exit(1)
 	}
-	return strings.TrimSpace(stdout.String()), nil
-}
-
-func gitPush() error {
-	var stdout, stderr bytes.Buffer
-	cmd := exec.Command("git", "push")
-	cmd.Stdout = &stdout
-	cmd.Stderr = &stderr
 
-	if err := cmd.Run(); err != nil {
-		return fmt.Errorf("failed to push: %s %s", stderr.String(), stdout.String())
-	}
-	return nil
-}
+	execArgs.execCommand = execArgs.userCommand
 
-func Exec(cmd *cobra.Command, args []string) error {
-	cmd.SilenceUsage = true
+	if !config.ExecAttach {
+		escaped := strings.ReplaceAll(strings.Join(execArgs.userCommand, " "), "\"", "\\\"")
 
-	dir, err := config.GetActiveProjectPath()
-	if err != nil {
-		ui.Errorf("Couldn't get active project path. Make sure you're in a project " +
-			"directory or supply a path: \n" + err.Error())
-		os.Exit(1)
+		execArgs.execCommand = []string{"nohup", "bash", "-c", "\"", escaped, "\""}
+		execArgs.execCommand = append(execArgs.execCommand, ">", "exec.log", "2>&1", "&", "echo", "$!", ">", "./pid.nohup", "&&", "sleep", "1")
 	}
 
-	if s, err := os.Stat(dir); err != nil || !s.IsDir() {
-		ui.Errorf("Couldn't find directory %q", dir)
-		os.Exit(1)
-	}
+	return execArgs
+}
 
-	ui.Infof("Gathering context from %q", dir)
-	buf := &bytes.Buffer{}
-	if err := gatherContext(dir, buf, "zip"); err != nil {
-		return err
+func (e *execCommandFlow) getExec(cmd *cobra.Command, execCmd execCmdArgs) (chan types.Exec, bool, chan error) {
+	ctx := cmd.Context()
+
+	sendError := func(err error) chan error {
+		errCh := make(chan error, 1)
+		errCh <- err
+		return errCh
 	}
 
-	// check config to see if the user has auto-commit enabled
-	// if auto-commit is set to ask, ask the user if they want to commit
-	// if yes, get git remote url, commit and push
+	ui.Infof("Initializing session...")
 
-	gitRemote, err := getGitRemoteURL()
+	execCh, errCh, err := execCreateAndWatch(ctx, types.ExecConfig{Command: execCmd.userCommand}, types.GitConfig{})
 	if err != nil {
-		return err
-	}
-	if err = gitAdd(); err != nil {
-		return err
+		return nil, false, sendError(err)
 	}
 
-	commitID, err := gitCommit()
-	if err != nil {
-		return err
-	}
-	if err = gitPush(); err != nil {
-		return err
-	}
-	commitURL := fmt.Sprintf("%s/commit/%s", gitRemote, commitID)
-	ui.Infof("Commit URL: %s", commitURL)
-
-	execConfig := types.ExecConfig{
-		Image:   "",
-		Command: args,
-		Keys:    nil,
-		Volumes: nil,
-		Src: &types.SourceContext{
-			MountPath: "/home/ubuntu",
-			Context:   io.NopCloser(buf),
-		},
-	}
+	const alwaysNewExec = true
 
-	gitConfig := types.GitConfig{
-		CommitID: &commitID,
-		GitURL:   &gitRemote,
-	}
+	return execCh, alwaysNewExec, errCh
+}
 
-	_, err = sessionCreate(cmd.Context(), execConfig, gitConfig)
-	if err != nil {
-		return err
-	}
+func (e *execCommandFlow) onTerminate(ctx context.Context, execID string) error {
+	ui.Infof("Session %q exited. Use 'unweave terminate' to stop the session.", execID)
```

Comment:
Since the exec is a command running in a container, it should automatically exit when the command exits right? So we shouldn't need this.

Reply:
At this point, we don't know if the ssh session terminated because the command was run detached, or the ssh session terminated because the command was running attached and finished. 

Another thought is that out exec logs currently exist inside the exec, and no where else. So if you want to read the logs after the command has completed, you need to keep the exec around. 

While this logs problem (and the above ssh problem) still exist, I recon we might just ask the user to terminate manually (maybe?). 

Both of these problems are solvable, but maybe we can ship this improvement first. :thinking: